### PR TITLE
Add attribute directives to the configuration files

### PR DESCRIPTION
**Pull request by wallento**
As suggested in #1514 this allows to set all comment attributes (`/* verilator <attribute> */`) in configuration files.
There are some minor things to do:

- [x] Check for proper semantics, currently I suppose that some invalid things could be done.
- [x] Add wildmatch for files for the case types
- [x] Cache wildcard matches as in the current ignore cases. It is not as critical as the new `apply*` functions are visited for specific AST node types, not all FileLines..

### CODE REVIEW

File: src/V3Config.cpp [hunk 1]
```diff
@@ -145,9 +146,289 @@ class V3ConfigIgnores {
...
+    void addVarAttr(FileLine* fl, const string& module, const string& ftask, const string& var,
+                    AstAttrType type, AstSenTree* sensep) {
+        m_varAttrs[module][ftask][var].push_back(std::make_tuple(type, fl, sensep));
+    }
+    void addFunctionIsolate(const string& module, const string& function) {
+        m_functionIsolate[module].insert(function);
+    }
+    void applyVarAttr(AstNodeModule* modulep, AstVar* varp) {
+        const string module = modulep->name();
+        const string var = varp->name();
```
Comment (wsnyder) [inline src/V3Config.cpp]: Codacy Issue found: [Variable 'var' is assigned a value that is never used.](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4742436)

**Comment by wsnyder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4742436&bid=15826673)

File: src/V3Config.cpp [hunk 2]
```diff
@@ -145,9 +146,289 @@ class V3ConfigIgnores {
...
+    inline static V3ConfigVarAttr& singleton() { return s_singleton; }
+    void addVarAttr(FileLine* fl, const string& module, const string& ftask, const string& var,
+                    AstAttrType type, AstSenTree* sensep) {
+        m_varAttrs[module][ftask][var].push_back(std::make_tuple(type, fl, sensep));
+    }
+    void addFunctionIsolate(const string& module, const string& function) {
+        m_functionIsolate[module].insert(function);
+    }
+    void applyVarAttr(AstNodeModule* modulep, AstVar* varp) {
+        const string module = modulep->name();
```
Comment (wsnyder) [inline src/V3Config.cpp]: Codacy Issue found: [Variable 'module' is assigned a value that is never used.](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4742436)

File: src/V3Config.cpp [hunk 3]
```diff
@@ -145,9 +146,311 @@ class V3ConfigIgnores {
 
 V3ConfigIgnores V3ConfigIgnores::s_singleton;
 
+//######################################################################
+// variable attributes (public, clock, isolate_assignments, ...)
+
+class V3ConfigVarAttr {
+    struct AttrEntry {
+        AstAttrType type;
+        FileLine*   flp; // currently don't use it, but may be needed for error reporting later
```
Comment (wsnyder) [inline src/V3Config.cpp]: Codacy Issue found: [struct member 'AttrEntry::flp' is never used.](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4742436)

**Comment by wsnyder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4742436&bid=15826673)

**Comment by wallento** [comment]
I am kind of happy now with how it is. Tests adopted for proper coverage of the feature.

About the caching I am undecided. In terms of runtime it would be better to have each of the config classes to maintain full expanded maps of files->task->vars->attr (worst, but common case). This can explode in terms of memory usage and I am not sure if caching would actually be that much of an improvement. Maybe revisit later?

File: bin/verilator [hunk 4]
```diff
@@ -2793,6 +2793,137 @@ and range of line numbers (or all lines if omitted).
...
+or vluint64_t.  This may be useful if the port width is parameterized and
+different of such modules interface a templated module (such as a transactor)
+or for other reasons.  In general you should avoid using this attribute when
+not necessary as with increasing usage of sc_bv the performance decreases
+significantly.
+
+Same as /*verilator sc_bv*/, see L</"LANGUAGE EXTENSIONS"> for more
+information.
+
+=item sc_clock -module "<modulename>" [-task "<taskname>"] -signal "<signame>"
```
Comment (wsnyder) [inline bin/verilator]: We can remove sc_clock in this as SC 1.2 is very old now and AFAIK unused.

File: bin/verilator [hunk 5]
```diff
@@ -2928,7 +3059,8 @@ per the C standard (it's unspecified in Verilog).
 
 Specifies the entire begin/end block should be ignored for coverage
 analysis.  Must be inside a basic block, e.g. within a begin/end pair.
-Same as /* verilator coverage_block_off */.
+Same as /* verilator coverage_block_off */ and coverage_block_off in
+configuration files.
```
Comment (wsnyder) [inline bin/verilator]: "... and C<coverage_block_off> in ....

(So the font makes it clear this is a command. Ditto others.)

File: src/V3Config.cpp [hunk 6]
```diff
@@ -145,9 +146,310 @@ class V3ConfigIgnores {
...
+    void addFunctionIsolate(const string& module, const string& function) {
+        m_functionIsolate[module].insert(function);
+    }
+
+    void applyVarAttr(const AstNodeModule* modulep, const AstNodeFTask* ftaskp, AstVar* varp) {
+        const string& module = modulep->name();
+        const string& ftask = (ftaskp ? ftaskp->name() :  "");
+
+        for (AttrMap::const_iterator it = m_varAttrs.begin(); it != m_varAttrs.end(); ++it) {
+            if (VString::wildmatch(module.c_str(), it->first.c_str())) {
```
Comment (wsnyder) [inline src/V3Config.cpp]: If wildmatch doesn't have a flavor that takes two std strings please add one, then remove c_str()s here.

File: src/V3Config.cpp [hunk 7]
```diff
@@ -145,9 +146,310 @@ class V3ConfigIgnores {
...
+                applyVarAttrForModule(it->second, ftask, varp);
+            }
+        }
+    }
+    void applyFunctionIsolate(const AstNodeModule* modulep, AstNodeFTask* funcp) {
+        std::map<string,FunctionList>::const_iterator it;
```
Comment (wsnyder) [inline src/V3Config.cpp]: Please use a typedefed::iterator

File: src/V3Config.cpp [hunk 8]
```diff
@@ -145,9 +146,310 @@ class V3ConfigIgnores {
...
+
+    void applyCase(AstCase* nodep) {
+        const string& file = nodep->fileline()->filename();
+        int lineno = nodep->fileline()->lineno();
+
+        for (CaseList::const_iterator it = m_parallels.begin();
+             it != m_parallels.end(); ++it) {
+            if (VString::wildmatch(file.c_str(), it->first.c_str())) {
+                if ((m_parallels[file].find(0) != m_parallels[file].end()) ||
+                    (m_parallels[file].find(lineno) != m_parallels[file].end())) {
```
Comment (wsnyder) [inline src/V3Config.cpp]: Perhaps should have standard class to save file/line and match them?

In general seems a lot if code for each pragma.  Might be fine but see if that's the best we want to do.

Nit || goes first on line.

File: src/V3EmitXml.cpp [hunk 9]
```diff
@@ -114,6 +114,7 @@ class EmitXmlFileVisitor : public AstNVisitor {
         puts(" origName="); putsQuoted(nodep->origName());
         if (nodep->level()==1 || nodep->level()==2)  // ==2 because we don't add wrapper when in XML mode
             puts(" topModule=\"1\"");  // IEEE vpiTopModule
+        if (nodep->modPublic()) puts(" public=\"true\"");
```
Comment (wsnyder) [inline src/V3EmitXml.cpp]: Suggest pull these xml improvements into a separate pull request. Once regressions pass you can squash and push to master.

File: src/V3LinkParse.cpp [hunk 10]
```diff
@@ -106,6 +108,10 @@ class LinkParseVisitor : public AstNVisitor {
 
     // VISITs
     virtual void visit(AstNodeFTask* nodep) {
+#ifndef _V3ERROR_NO_GLOBAL_
+        V3Config::applyFTask(m_modp, nodep);
+#endif
+
```
Comment (wsnyder) [inline src/V3LinkParse.cpp]: These ifndefs should not be needed.

File: src/verilog.l [hunk 11]
```diff
@@ -135,16 +135,44 @@ vnum    {vnum1}|{vnum2}|{vnum3}|{vnum4}|{vnum5}
...
+  "sformat"             { FL; return yVLT_SFORMAT; }
   "tracing_off"         { FL; return yVLT_TRACING_OFF; }
   "tracing_on"          { FL; return yVLT_TRACING_ON; }
 
+  -?"-block"            { FL; return yVLT_D_BLOCK; }
   -?"-file"             { FL; return yVLT_D_FILE; }
+  -?"-function"         { FL; return yVLT_D_FUNCTION; }
   -?"-lines"            { FL; return yVLT_D_LINES; }
+  -?"-module"           { FL; return yVLT_D_MODULE; }
   -?"-msg"              { FL; return yVLT_D_MSG; }
```
Comment (wsnyder) [inline src/verilog.l]: Thought we were  going to deprecate --msg for --rule?

File: test_regress/t/t_clk_concat.vlt [hunk 12]
```diff
@@ -0,0 +1,8 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// This file ONLY is placed into the Public Domain, for any use,
+// without warranty, 2019 by Stefan Wallentowitz.
+
+`verilator_config
+
+clocker -module "t" -var "clk*"
```
Comment (wsnyder) [inline test_regress/t/t_clk_concat.vlt]: Note missing newline

File: src/V3EmitXml.cpp (hunk 9 above)
Comment (wallento) [reply src/V3EmitXml.cpp]: Thanks, see #2079

File: src/verilog.l (hunk 11 above)
Comment (wallento) [reply src/verilog.l]: This was running in parallel. I will create a separate PR for the rename and base this on top.

File: src/V3Config.cpp (hunk 8 above)
Comment (wallento) [reply src/V3Config.cpp]: Yeah, I believe there is some potential to consolidate, especially once the code for caching is added.

**Comment by wallento** [comment]
Next iteration, amended as suggested, the caching will hopefully consolidate some of the stuff, especially as the (module,task,var) triplets and (file,lineno) are the base for all of those..

**Comment by wallento** [comment]
I have refactored this. It is best to view the file at this specific version, I just want to keep the original one too for now as comparison. This seems a bit over-engineered now, good is that I have a generic wildcard resolution class now, that also caches lookups. I am not sure if we really need it full blown as LinkParse visits it pretty straightforwardly. But it may make sense to rework it to not store resolved entities.

But overall I am also more happy with this solution now. It does not make much sense to merge the ignore stuff into it, as they work differently.

**Comment by wallento** [comment]
Please ignore the naming, style etc. It is very preliminary, I just want to explore how to optimize the implementation.

**Comment by wallento** [comment]
Another update and squashed into one. The original code can be found here: https://github.com/wallento/verilator/tree/config-old

The major issue I could anticipate is that it creates a resolved node for every module, ftask and var, which can be a memory issue. Taken that the number of configuration switches is probably much smaller, I think its best to rework it to work on pointers and not references and return NULL if an entity was not matched instead of creating a default one.

Right direction, @wsnyder?

**Comment by wsnyder** [comment]
Sorry I'm not sure which commit/file/line you are referring to when you asked about pointers.

**Comment by wallento** [comment]
Thats the current working state: https://github.com/verilator/verilator/pull/2072/commits/d6fca70669dcaf0580f35cc08bb48a72c5e80ea2

I fell in love with references yet again and it has some nice lines like this one: https://github.com/verilator/verilator/pull/2072/commits/d6fca70669dcaf0580f35cc08bb48a72c5e80ea2#diff-294974d3de15caf77ef41767fb5a4ffcR485

The problem with this is that it creates a "default" object in the resolver on each stage if it didn't match. Not sure what we are talking about, but even a couple of Byte per variable could easily sum up I assume.

It would be better to instead return pointers on non-matches and

```
ConfigModule* mod = V3ConfigAttributes::singleton().modules().resolve(modulep->name());
if (!mod) return;
ConfigFTask* ftask = mod..ftasks().resolve(ftaskp->name());
if (!ftask) return;
ftask.apply(ftaskp);
```

More code, but less runtime memory footprint.

**Comment by wallento** [comment]
I reworked it to this now. When there are no config file settings, there will be just two objects of a couple of Byte and a few instructions. As the tree of wildcard resolves is sparse now the memory footprint should not be an issue, another addition could be to only cache the last resolved name. This will always hit on module and ftask due to how LinkParse iterates and it will never access a module or FTask again then, I believe. Only for waive (see other PR based on this one) there is arguably an advantage, which may be negligible.

The following VLT settings are interesting data points to check on a very large design:

```
public -module "*" -var "*"
public -module "*" -task "*" -var "*"
```

and

```
public -module "*"* -var "*clk*"
public -module "*"* -task "*" -var "*clk*"
```

Impact of this PR on runtime and memory footprint would be of interest. I can have a look, but only have rather small designs I suppose at hand.

**Comment by wallento** [comment]
Merged the old code into the new resolver logic, code cleanup. My tests show there is no significant difference in terms of runtime and memory usage for wildcards that match everything.

I think it slowly converges..

**Comment by wallento** [comment]
Minor fixes pushed, ready for review from my point of view.

File: src/V3String.cpp [hunk 13]
```diff
@@ -74,6 +74,15 @@ bool VString::wildmatch(const char* s, const char* p) {
...
 
+bool VString::wildmatch(const string& s, const string& p) {
+    return wildmatch(s.c_str(), p.c_str());
+}
+
+bool VString::isWildcard(const string &p) {
+    return ((p.find("*") != string::npos) || (p.find("?") != string::npos));
+}
+
+
```
Comment (wsnyder) [inline src/V3String.cpp]: Extra newline

File: src/verilog.y [hunk 14]
```diff
@@ -2460,6 +2482,11 @@ cellpinItemE<pinp>:		// IEEE: named_port_connection + empty
 //************************************************
 // EventControl lists
 
+attr_event_controlE<sentreep>:
+		/* empty */			{ $$ = NULL; }
+	|	attr_event_control	{ $$ = $1; }
```
Comment (wsnyder) [inline src/verilog.y]: For all insertions please tab the {'s out to the standard 7th tab column.
For really long lines the { goes on a new line with 3 tabs.

File: src/verilog.y [hunk 15]
```diff
@@ -5590,6 +5617,15 @@ vltItem:
...
 	|	vltOnFront yVLT_D_FILE yaSTRING yVLT_D_LINES yaINTNUM '-' yaINTNUM	{ V3Config::addIgnore($1,true,*$3,$5->toUInt(),$7->toUInt()+1); }
+	|	vltVarAttrFront vltDModuleE vltDFTaskE vltVarAttrVarE attr_event_controlE { V3Config::addVarAttr($<fl>1,*$2,*$3,*$4,$1,$5); }
+	|	vltInlineFront vltDModuleE vltDFTaskE { V3Config::addInline($<fl>1,*$2,*$3,$1); }
+	|	yVLT_COVERAGE_BLOCK_OFF yVLT_D_FILE yaSTRING { V3Config::addCoverageBlockOff(*$3,0); }
+	|	yVLT_COVERAGE_BLOCK_OFF yVLT_D_FILE yaSTRING yVLT_D_LINES yaINTNUM { V3Config::addCoverageBlockOff(*$3,$5->toUInt()); }
+	|	yVLT_COVERAGE_BLOCK_OFF yVLT_D_MODULE yaSTRING yVLT_D_BLOCK yaSTRING { V3Config::addCoverageBlockOff(*$3,*$5); }
+	|	yVLT_FULL_CASE yVLT_D_FILE yaSTRING { V3Config::addCaseFull(*$3,0); }
+	|	yVLT_FULL_CASE yVLT_D_FILE yaSTRING yVLT_D_LINES yaINTNUM { V3Config::addCaseFull(*$3,$5->toUInt()); }
+	|	yVLT_PARALLEL_CASE yVLT_D_FILE yaSTRING { V3Config::addCaseParallel(*$3,0); }
+	|	yVLT_PARALLEL_CASE yVLT_D_FILE yaSTRING yVLT_D_LINES yaINTNUM { V3Config::addCaseParallel(*$3,$5->toUInt()); }
```
Comment (wsnyder) [inline src/verilog.y]: Rather than { far to the right these probably would look better like this:

        |       vltVarAttrFront vltDModuleE vltDFTaskE vltVarAttrVarE attr_event_controlE
                        { V3Config::addVarAttr($<fl>1,*$2,*$3,*$4,$1,$5); }
        |       vltInlineFront vltDModuleE vltDFTaskE
                        { V3Config::addInline($<fl>1,*$2,*$3,$1); }

File: src/verilog.y [hunk 16]
```diff
@@ -5618,6 +5654,40 @@ vltOnFront<errcodeen>:
 			  if ($$ == V3ErrorCode::EC_ERROR) { $1->v3error("Unknown Error Code: "<<*$3<<endl);  } }
 	;
 
+vltDModuleE<strp>:
+		/* empty */				{ static string unit="__024unit"; $$=&unit; }
```
Comment (wsnyder) [inline src/verilog.y]: Please space around =.  (Know this was probably cut-and-pasted from something old.)

File: test_regress/t/t_assert_synth_full.pl [hunk 17]
```diff
@@ -21,7 +21,7 @@
     check_finished => 0,
     fails => $Self->{vlt_all},
     expect =>
-'%Error: t_assert_synth.v:\d+: Assertion failed in top.t: synthesis full_case'
+'%Error: t_assert_synth.v:30: Assertion failed in top.t: synthesis full_case'
```
Comment (wsnyder) [inline test_regress/t/t_assert_synth_full.pl]: If it's critical to prove the test adding the line number is ok, otherwise better to leave it general as if the test changes we then need to edit this .pl file.

File: test_regress/t/t_vlt_syntax_bad.v [hunk 18]
```diff
@@ -0,0 +1,2 @@
+module t;
+endmodule
```
Comment (wsnyder) [inline test_regress/t/t_vlt_syntax_bad.v]: Please trailing newlines; I just committed a test change to fail that.

Comment (wallento) [reply test_regress/t/t_vlt_syntax_bad.v]: Sorry, recently switched editors, I think its configured again but good to have a test :)

File: src/verilog.y (hunk 15 above)
Comment (wallento) [reply src/verilog.y]: Changed it for all rules in this item for consistency

Comment (wallento) [reply src/verilog.y]: Also more room to add spaces to the arguments then :)

File: src/V3Config.cpp [hunk 19]
```diff
@@ -49,114 +242,241 @@ class V3ConfigLine {
...
+        // Apply to this case at this line
+        int lineno = nodep->fileline()->lineno();
+        if (lineMatch(lineno, FULL_CASE)) nodep->fullPragma(true);
+        if (lineMatch(lineno, PARALLEL_CASE)) nodep->parallelPragma(true);
     }
     inline void applyIgnores(FileLine* filelinep) {
         // HOT routine, called each parsed token line
-        if (m_lastLineno != filelinep->lastLineno()
-            || m_lastFilename != filelinep->filename()) {
+        if (m_lastIgnore.lineno != filelinep->lineno()) {
```
[the remark this answers is not in the document]
Comment (wallento) [reply src/V3Config.cpp]: Each filename has it's own context now as this is accessed via the WildcardResolver. m_lastfilename logic moved to that as the absBuild did. Added a comment.

[the remark this answers is not in the document]
Comment (wallento) [reply src/V3Config.cpp]: It will also only call this function if the name resolved.

[the remark this answers is not in the document]
Comment (wallento) [reply src/V3Config.cpp]: Probably this is not inlined, but doesn't have to. Instead maybe the resolve function, right? This is "hotter".

[the remark this answers is not in the document]
Comment (wsnyder) [reply src/V3Config.cpp]: I don't thonk it is so hot we need to get inline-fancy.

File: test_regress/t/t_assert_synth_full.pl (hunk 17 above)
Comment (wsnyder) [reply test_regress/t/t_assert_synth_full.pl]: Perhaps a .out/golden file is preferrable?  Easier to update.

File: src/V3Config.cpp [hunk 20]
```diff
@@ -49,114 +242,241 @@ class V3ConfigLine {
...
-class V3ConfigIgnores {
-    typedef std::multiset<V3ConfigLine> IgnLines;  // list of {line,code,on}
-    typedef std::map<string,IgnLines> IgnFiles;  // {filename} => list of {line,code,on}
+// Some attributes are attached to entities of the occur on a fileline
+typedef enum {
+    COVERAGE_BLOCK_OFF,
+    FULL_CASE,
+    PARALLEL_CASE,
+    ENUM_SIZE
+} V3ConfigLineAttributesType;
```
[the remark this answers is not in the document]
Comment (wallento) [reply src/V3Config.cpp]: I wanted to keep the bitfields small and not use sparse fields. On the other hand I assume its not a major issue, what do you think?

[the remark this answers is not in the document]
Comment (wsnyder) [reply src/V3Config.cpp]: I'd vote for easier maintenance (shared enum) in this case.

File: src/V3LinkParse.cpp [hunk 21]
```diff
@@ -189,6 +193,9 @@ class LinkParseVisitor : public AstNVisitor {
             return;
         }
 
+        // Maybe this variable has a signal attribute
+        V3Config::applyVarAttr(m_modp, (m_funcp ? m_funcp : m_ftaskp), nodep);
```
[the remark this answers is not in the document]
Comment (wallento) [reply src/V3LinkParse.cpp]: Ah, right, some leftover..

File: src/V3Config.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (wallento) [reply src/V3Config.cpp]: The easier maintenance has it

**Comment by wallento** [comment]
Thanks for the review, amended with your suggestions and force pushed.

> I think you're missing tests for no_clocker.

Yes, indeed, I didn't find a test for the comment attribute and was not really sure how to test it and ultimately forgot. I will add this asap.

**Comment by wallento** [comment]
no_clocker test included now

**Comment by wallento** [comment]
Added to `Changes` too
